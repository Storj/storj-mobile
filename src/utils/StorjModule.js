import {
    NativeModules,
    Platform
} from 'react-native'
import BucketModel from '../models/BucketModel';
import FileModel from '../models//FileModel';
import keysModel from '../models/keysModel';

//TODO: StorjModule wil send us only Response objects, 
// so all try/catch blocks should be removed and all error logic should be after checking !isSuccess
const StorjLib = (() => {
    let instance = null;

    const storjLibAndroid = NativeModules.StorjLibAndroid;
    const storjLibIos = NativeModules.StorjLibIos;
    const isAndroid = Platform.OS === 'android';

    const storjLib = isAndroid ? storjLibAndroid : storjLibIos;

    /**
     * This module wraps Native Modules for StorjLib.
     * Will be rafactored after iteration 2.
     * We should implement more functionality to have more deep understanding of needed arcitecture
     */
    class StorjModule {

        getStorjLibNativeModule(){
            return storjLib;
        }

        /**
         * Generate mnemonic
         * @returns {Promise<boolean>}
         */
        async generateMnemonic() {
            try {
                let response = await storjLib.generateMnemonic();

                if(!response.isSuccess){
                    console.log('generateMnemonic ', response.error.message);
                }

                return response;
            } catch(e) {
                console.log(e);
            }
        }

        /**
         * Check if mnemonic provided has valid format
         * @param {string} mnemonic
         * @returns {Promise<boolean>} 
         */
        async checkMnemonic(mnemonic) {    
            let response = await storjLib.checkMnemonic(mnemonic);

            if(!response.isSuccess){
                console.log('checkMnemonic ', response.error.message);
            }

            return response.isSuccess;
        };
    
        /**
         * Send new registration request
         * @param {string} email 
         * @param {string} password 
         * @param {function} sucessCallback Callback that is called on Success, returns email address from the request
         * @param {function} errorCallback Callback for error handeling, returns error message
         */
        async register(email, password, errorCallback) {
            let response = await storjLib.register(email, password);

            if(!response.isSuccess){
                console.log('register ', response.error.message);
            }

            return response;
        };
    
        /**
         * Verify if user exist in storj network
         * @param {string} email 
         * @param {string} password 
         * @returns {Promise<boolean>}
         */
        async verifyKeys(email, password) {            
            let response = await storjLib.verifyKeys(email, password);

            if(!response.isSuccess){
                console.log('verifyKeys ', response.error.message);
            }

            return response.isSuccess;
        };
    
        /**
         * Check if auth file allready exist on the device
         * @returns {Promise<boolean>}
         */
        async keysExists() {   
            let response = await storjLib.keysExists();

            if(!response.isSuccess){
                console.log('keysExists ', response.error.message);
            }

            return response.isSuccess;
        };
    
        /**
         * Creates new auth file for given credentials and stores it on the device
         * and saves them in the current context
         * @param {string} email 
         * @param {string} password 
         * @param {string} mnemonic 
         * @param {string} passcode optional, pass if you want to protect auth file with additional password
         * @returns {Promise<boolean>}
         */
        async importKeys(email, password, mnemonic, passcode) {            
            let response = await storjLib.importKeys(email, password, mnemonic, /*passcode*/"");

            if(!response.isSuccess){
                console.log('importKeys ', response.error.message);
            }

            return response.isSuccess;
        };
    
        /**
         * 
         * @param {string} passcode needed if user has protected your auth file with additional password
         */
        async getKeys(passcode) {
            let response = await storjLib.getKeys(/*passcode*/"");

            if(!response.isSuccess) {
                console.log('getKeys ', response.error.message);
            }

            return response;
        };
    
        /**
         * List buckets for logged in user
         * @returns {Promise<BucketModel[]>}
         */
        async getBuckets() {
            let result = [];
            let response = await storjLib.getBuckets();

            if(!response.isSuccess) {
                console.log('getBuckets ', response.error.message);
                return result;
            }

            result = JSON.parse(response.result).map((bucket) => {
                return new BucketModel(bucket);
            });

            return result;
        }

        /**
         * Deletes bucket by Id
         * @param {string} bucketId 
         */
        async deleteBucket(bucketId) {
            let response = await storjLib.deleteBucket(bucketId);

            if(!response.isSuccess) {
                console.log('deleteBucket ', response.error.message);
            }

            return response;
        }

        /**
         * download file to storj network
         * @returns {Promise<any>}
         */
        async downloadFile(bucketId, fileId, localPath) {
            let response = await storjLib.downloadFile(bucketId, fileId, localPath);

            if(!response.isSuccess) {
                console.log('downloadFile ', response.error.message);
            }

            return response;
        }

        /**
         * cancel file downloading
         * @returns {Promise<any>}
         */
        async cancelDownload(fileRef) {
            let response = await storjLib.cancelDownload(fileRef);

            if(!response.isSuccess) {
                console.log('cancelDownload ', response.error.message);
            }

            return response;
        }

        /**
         * cancel file uploading
         * @returns {Promise<any>}
         */
        async cancelUpload(fileRef) {
            let response = await storjLib.cancelUpload(fileRef);

            if(!response.isSuccess) {
                console.log('cancelUpload ', response.error.message);
            }

            return response;
        }

        /**
         * download file to storj network
         * @returns {Promise<any>}
         */
         async uploadFile(bucketId, localPath) {
            let response = await storjLib.uploadFile(bucketId, localPath);

            if(response.isSuccess) {
                response.result = new FileModel(JSON.parse(response.result));
            } else {
                console.log('uploadFile ', response.error.message);
            }

            return response;
        };

        /**
         * List buckets for logged in user
         * @returns {Promise<FileModel[]>}
         */
        async listFiles(bucketId) {
            let response = await storjLib.listFiles(bucketId);

            if(response.isSuccess) {
                response.result = JSON.parse(response.result).map(file => {
                    return new FileModel(file);
                });      
            } else {
                console.log('listFiles ', response.error.message);
            }

            return response;
        };
    
        /**
         * Create bucket
         * @returns {Promise<BucketModel>}
         */
        async createBucket(bucketName) {
            let response = await storjLib.createBucket(bucketName);
            
            if(response.isSuccess) {
                response.result = new BucketModel(JSON.parse(response.result));
            } else {
                console.log('createBucket ', response.error.message);
            }

            return response;         
        };

        /**
         * 
         * @param {string} bucketId 
         * @param {string} fileId 
         */
        async deleteFile(bucketId, fileId) {
            let response = await storjLib.deleteFile(bucketId, fileId);

            if(!response.isSuccess) {
                console.log('deleteFile ', response.error.message);
            } 

            return response;
        }
    }

    return {
        /**
        * @returns {StorjModule}
        */
        getInstance: function() {
            if(!instance) {
                instance = new StorjModule();
            }

            return instance;
        }
    };
})();

export default StorjLib.getInstance();


