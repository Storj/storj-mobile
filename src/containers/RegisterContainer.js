import PropTypes from 'prop-types';
import { AsyncStorage } from 'react-native';
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { registerActionsCreators } from '../reducers/authentification/authActions';
import { RegisterStateModel } from '../models/RegisterStateModel';
import { RegisterErrorModel } from '../models/RegisterErrorModel';
import RegisterComponent from '../components/RegisterComponent';
import StorjModule from '../utils/StorjModule';
import validator from '../utils/validator';
import infoScreensConstants from '../utils/constants/infoScreensConstants';

const FIRST_ACTION = 'FIRST_ACTION';

/**
 * Redux container for register component
 */
export class RegisterContainer extends Component {
	constructor(props) {
        super(props);

        this.state = {
            stateModel: new RegisterStateModel(),
            errorModel: new RegisterErrorModel(),
            isLoading: false
        };
        
    };
    
    /**
     * Remove react-navigation stackNavigator page headers
     */
    static navigationOptions = {
        header: null
    };

    /**
     * Changing internal state when user login inputting
     * @param {*} value current value in Input
     */
    onChangeEmail(value) {
        this.setState({
            stateModel: 
                new RegisterStateModel(
                    value, 
                    this.state.stateModel.password, 
                    this.state.stateModel.passwordRepeat, 
                    this.state.stateModel.areTermsAccepted)
        });
    };

    /**
     * Changing internal state when user inputting password
     * @param {*} value current value in Input
     */
    onChangePassword(value) {
        this.setState({
            stateModel: 
            new RegisterStateModel(
                this.state.stateModel.email, 
                value, 
                this.state.stateModel.passwordRepeat, 
                this.state.stateModel.areTermsAccepted)
            });
    };

    /**
     * Changing internal state when user repeating password
     * @param {*} value current value in Input
     */
    onChangePasswordRepeat(value) {
        this.setState({
            stateModel: 
                new RegisterStateModel(
                    this.state.stateModel.email, 
                    this.state.stateModel.password, 
                    value, 
                    this.state.stateModel.areTermsAccepted)
            });
    };

    /**
     * Changing internal state when user selects terms of use checkbox
     * @param {bool} value current value in Input
     */
    onChangeTermsAcceptence(value) {
        this.setState({
            stateModel: 
                new RegisterStateModel(
                    this.state.stateModel.email, 
                    this.state.stateModel.password, 
                    this.state.stateModel.passwordRepeat, 
                    value)
            });
    };
    
    /**
     * Handle if was allready in use
     */
    async handleFirstLaunch() {
        if(!await AsyncStorage.getItem(FIRST_ACTION)) 
            await AsyncStorage.setItem(FIRST_ACTION, 'true');
        
    };

    /**
     * Submit registration info
     */
    async onSubmit() {
        if(this.state.isLoading) return;
        
        this.setState({ isLoading: true });
        let isEmailValid = validator.isEmail(this.state.stateModel.email);
        let isPasswordValid = this.state.stateModel.password;
        let isPasswordMatch = this.state.stateModel.password === this.state.stateModel.passwordRepeat;

        if(isEmailValid && isPasswordValid && isPasswordMatch 
           && this.state.stateModel.areTermsAccepted) {
               await this.register();          
        } else {
            this.setState({
                errorModel: new RegisterErrorModel(
                    !isEmailValid,
                    !isPasswordValid,
                    !isPasswordMatch,
                    !this.state.stateModel.areTermsAccepted)
            });
        }

        this.setState({
            isLoading: false
        });
    };

    /**
    * Preparing data and calling StorjLib register function
    */
    async register() {
        this.props.register(this.state.stateModel.email, this.state.stateModel.password);

        const registrationResult = await StorjModule.register(
            this.state.stateModel.email, 
            this.state.stateModel.password, 
            this.registerErrorCallback.bind(this));


        if(!registrationResult.isSuccess) {
            const errorMessage = registrationResult.errorMessage;
            this.props.registerError(errorMessage);

            //TODO: improve in future, we need better error codes from StorjLib
            if(errorMessage.includes('is already registered')) { 
                this.props.redirectToAuthFailureScreen({ 
                    mainText: infoScreensConstants.registerFailureMainText, 
                    additionalText: infoScreensConstants.registerFailureAdditionalText 
                });
            }    
            
            return;
        }
        
        this.props.registerSuccess(registrationResult.mnemonic);
        this.handleFirstLaunch();
        this.props.redirectToRegisterSuccessScreen(registrationResult.mnemonic, this.state.stateModel.email);    
    };

    /**
    * Error callback for register method from StorjLib.
    */
    registerErrorCallback(error) {  
        this.setState({ isLoading: false });
        this.props.registerError(error.message);
    
        this.props.redirectToAuthFailureScreen({ 
            mainText: 'Something Went Wrong', 
            additionalText: error.message
        });
    };

    /**
    * Navigate back to previous screen
    */
    navigateBack() {
        this.props.navigateBack();
    };

    /**
     * Navigate to TermsOfUseScreen
     */
    redirectToTermsOfUse() {
        this.props.redirectToTermsOfUse();
    }

    /**
    * Navigate to LoginScreen
    */
    redirectToLoginScreen() {
        this.props.redirectToLoginScreen();
    };

	render() {
		return(
                <RegisterComponent
                    isLoading = { this.state.isLoading }
                    onSubmit = { this.onSubmit.bind(this) }
                    onChangeEmail = { this.onChangeEmail.bind(this) }
                    onChangePassword = { this.onChangePassword.bind(this) }
                    onChangePasswordRepeat = { this.onChangePasswordRepeat.bind(this) }
                    onChangeTermsAcceptence = { this.onChangeTermsAcceptence.bind(this) }
                    navigateBack = { this.navigateBack.bind(this) }
                    redirectToTermsOfUse = { this.redirectToTermsOfUse.bind(this) }
                    redirectToLoginScreen = { this.redirectToLoginScreen.bind(this) }
                    isEmailError = { this.state.errorModel.isEmailError }
                    isPasswordError = { this.state.errorModel.isPasswordError }
                    isPasswordMatchError = { this.state.errorModel.isPasswordMatchError }
                    isTermsAcceptedError = { this.state.errorModel.isTermsAcceptedError } />
		);
	};
}

/**
* connecting state and action creators to component props 
*/
function mapStateToProps(state) { return { user: state.authReducer.user }; };
function mapDispatchToProps(dispatch) { return bindActionCreators(registerActionsCreators, dispatch); };

/**
* Creating RegisterContainer mappings
*/
export default connect(mapStateToProps, mapDispatchToProps)(RegisterContainer);

/**
* Checking RegisterContainer correct prop types
*/
RegisterContainer.propTypes = {
    user: PropTypes.shape({
        isLoggedIn: PropTypes.bool,
        email: PropTypes.string,
        password: PropTypes.string,
        mnemonic: PropTypes.string,
        isLoading: PropTypes.bool,
        error: PropTypes.string
    })
};
