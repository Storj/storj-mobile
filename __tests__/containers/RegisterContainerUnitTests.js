import 'react-native';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import sinon from 'sinon';
import RegisterContainer from '../../src/containers/RegisterContainer';
import StorjModule from '../../src/utils/storjModule';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();
const initialState = { authReducer: { email: null, password: null, isSuccess: false, error: null, isLoading: false } };
const store = mockStore(initialState);

jest.mock('../../src/utils/StorjModule', () => {
    return {
        register: jest.fn((email, password) => {
            if(email === 'example@email.com' && password === '12345') {
                return {
                    isSuccess: true,
                    mnemonic: 'mnemonic',
                    errorMessage: null
                };
            } 
        })
    };
});

describe('RegisterContainer', ()=> {
    const numOfRegisterInvokes = 2;
    
    it('renders correctly', () => {

        const wrapper = shallow(
            <RegisterContainer />,
            { context: { store: store } }
        );

        expect(wrapper.dive()).toMatchSnapshot();
    });    

    it('onSubmit valid', async () => {
        const email = 'example@email.com';
        const password = '12345';
        const passwordRepeat = '12345';
        const areTermsAccepted = true;

        const wrapper = shallow(
            <RegisterContainer />,
            { context: { store: store } }
        );

        const handle = wrapper.dive().instance();
        handle.setState({ stateModel: { email, password, passwordRepeat, areTermsAccepted }});
        
        await handle.onSubmit();
        
        expect(StorjModule.register.mock.calls[0][0]).toBe(email);
        expect(StorjModule.register.mock.calls[0][1]).toBe(password);
        expect(StorjModule.register).lastCalledWith(email, password, StorjModule.register.mock.calls[0][2]);
    });

    it('trigers handleFirstLaunch on successful registration', async () => {
        const email = 'example@email.com';
        const password = '12345';
        const passwordRepeat = '12345';
        const areTermsAccepted = true;

        const wrapper = shallow(
            <RegisterContainer />,
            { context: { store: store } }
        );

        const handle = wrapper.dive().instance();
        handle.setState({stateModel: { email, password, passwordRepeat, areTermsAccepted }});

        const handleFirstLaunchSpy = sinon.spy(handle, 'handleFirstLaunch');
        
        await handle.onSubmit();

        expect(handleFirstLaunchSpy.calledOnce).toBe(true);
    });    

    it('onSubmit no password', async () => {
        const email = 'example@email.com';
        const password = null;
        const passwordRepeat = '12345';
        const areTermsAccepted = true;

        const wrapper = shallow(
            <RegisterContainer />,
            { context: { store: store } }
        );

        const handle = wrapper.dive().instance();
        handle.setState({stateModel: { email, password, passwordRepeat, areTermsAccepted }});
        
        await handle.onSubmit();
        
        expect(StorjModule.register.mock.calls.length).toBe(numOfRegisterInvokes);
        expect(handle.state.errorModel.isPasswordError).toBe(true);
    });

    it('onSubmit passwords dont match', async () => {
        const email = 'example@email.com';
        const password = '12345';
        const passwordRepeat = '1234';
        const areTermsAccepted = true;

        const wrapper = shallow(
            <RegisterContainer />,
            { context: { store: store } }
        );

        const handle = wrapper.dive().instance();
        handle.setState({stateModel: { email, password, passwordRepeat, areTermsAccepted }});
        
        await handle.onSubmit();
        
        expect(StorjModule.register.mock.calls.length).toBe(numOfRegisterInvokes);
        expect(handle.state.errorModel.isPasswordMatchError).toBe(true);
    });

    it('onSubmit terms are not accepted', async () => {
        const email = 'example@email.com';
        const password = '12345';
        const passwordRepeat = '12345';
        const areTermsAccepted = false;

        const wrapper = shallow(
            <RegisterContainer />,
            { context: { store: store } }
        );

        const handle = wrapper.dive().instance();
        handle.setState({stateModel: { email, password, passwordRepeat, areTermsAccepted }});
        
        await handle.onSubmit();
        
        expect(StorjModule.register.mock.calls.length).toBe(numOfRegisterInvokes);
        expect(handle.state.errorModel.isTermsAcceptedError).toBe(true);
    });

    it('onSubmit invalid email', async () => {
        const email = 'notValid email@.com';
        const password = '12345';
        const passwordRepeat = '12345';
        const areTermsAccepted = false;

        const wrapper = shallow(
            <RegisterContainer />,
            { context: { store: store } }
        );

        const handle = wrapper.dive().instance();
        handle.setState({stateModel: { email, password, passwordRepeat, areTermsAccepted }});
        
        await handle.onSubmit();
        
        expect(StorjModule.register.mock.calls.length).toBe(numOfRegisterInvokes);
        expect(handle.state.errorModel.isEmailError).toBe(true);
    });
});
