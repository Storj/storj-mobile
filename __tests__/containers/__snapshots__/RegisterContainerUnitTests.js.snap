// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterContainer renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RegisterContainer
    navigateBack={[Function]}
    redirectToAuthFailureScreen={[Function]}
    redirectToLoginScreen={[Function]}
    redirectToRegisterSuccessScreen={[Function]}
    redirectToTermsOfUse={[Function]}
    register={[Function]}
    registerError={[Function]}
    registerSuccess={[Function]}
    user={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "isEmailError": false,
      "isLoading": false,
      "isPasswordError": false,
      "isPasswordMatchError": false,
      "isTermsAcceptedError": false,
      "navigateBack": [Function],
      "onChangeEmail": [Function],
      "onChangePassword": [Function],
      "onChangePasswordRepeat": [Function],
      "onChangeTermsAcceptence": [Function],
      "onSubmit": [Function],
      "redirectToLoginScreen": [Function],
      "redirectToTermsOfUse": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "isEmailError": false,
        "isLoading": false,
        "isPasswordError": false,
        "isPasswordMatchError": false,
        "isTermsAcceptedError": false,
        "navigateBack": [Function],
        "onChangeEmail": [Function],
        "onChangePassword": [Function],
        "onChangePasswordRepeat": [Function],
        "onChangeTermsAcceptence": [Function],
        "onSubmit": [Function],
        "redirectToLoginScreen": [Function],
        "redirectToTermsOfUse": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
